### PR TITLE
Redox OS support

### PR DESCRIPTION
**Pull request by jackpot51**
This is a draft for feedback on the implementation. Features will be added and documentation improved, but for now, this is sufficient for all the examples to run on Redox.

- [x] Tested on all platforms changed
- [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [x] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [x] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

**Review by madsmtm — COMMENTED** [review]
I have not looked at the code in detail, but this is really cool @jackpot51!!

Are you on Matrix? ("@jeremy_soller"?) And would you be interested in joining the team and being the main maintainer for this platform?

**Comment by jackpot51** [comment]
Yes, I am on matrix @jeremy_soller, please consider me the maintainer of Redox OS support in winit.

**Comment by madsmtm** [comment]
Also linking previous PR: https://github.com/rust-windowing/winit/pull/759 ;)

**Comment by madsmtm** [comment]
If you could add yourself to `CODEOWNERS`, that'd be nice then.

I will gladly add you to the @rust-windowing organization, but I'll just let the other maintainers have their say first ;)

**Comment by jackpot51** [comment]
> If you could add yourself to `CODEOWNERS`, that'd be nice then.
Done

> I will gladly add you to the @rust-windowing organization, but I'll just let the other maintainers have their say first ;)
Okay, awesome

File: src/platform_impl/redox/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1300 @@
...
+const ORBITAL_FLAG_BORDERLESS: char = 'l';
+const ORBITAL_FLAG_RESIZABLE: char = 'r';
+const ORBITAL_FLAG_TRANSPARENT: char = 't';
+
+struct RedoxSocket {
+    fd: usize,
+}
+
+impl RedoxSocket {
+    unsafe fn open(path: &str) -> syscall::Result<Self> {
```
Comment (madsmtm) [inline src/platform_impl/redox/mod.rs:36]: I know that `winit` has not really historically been the safest library, but I would still like for new code to have more safety comments - would you mind adding that?

**Review by madsmtm — COMMENTED** [review]
You may want to add a section to the `README.md` about platform-specific usage btw

File: src/platform_impl/redox/mod.rs [hunk 2]
```diff
@@ -0,0 +1,1300 @@
...
+
+        // Writing a default TimeSpec will always trigger a time event
+        self.wake_socket
+            .write(&syscall::TimeSpec::default())
+            .unwrap();
+    }
+}
+
+unsafe impl Send for Window {}
+unsafe impl Sync for Window {}
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:1224]: I believe these can be removed now, they were an artifact from other members I had in there that did not impl Send/Sync but are now gone

File: src/platform_impl/redox/mod.rs (hunk 1 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:36]: Yes. I marked this unsafe because in order for RedoxSocket to be const while doing read and write, only specific paths can be supported that work as sockets. Opening a normal file would mean reads and writes would move the seek, for example, perhaps unexpectedly. But with sockets, each read and write is independent just like how UdpSocket can do recv_from/send_to without mut. Does this make sense or should I just mark this safe?

File: src/platform/mod.rs [hunk 3]
```diff
@@ -56,6 +56,7 @@ pub mod x11;
     target_os = "dragonfly",
     target_os = "freebsd",
     target_os = "netbsd",
-    target_os = "openbsd"
+    target_os = "openbsd",
+    target_os = "redox"
 ))]
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform/mod.rs:61]: Yep

**Comment by jackpot51** [comment]
This is ready for review. The test failures don't appear to be from changes I have made.

File: src/platform_impl/redox/mod.rs [hunk 4]
```diff
@@ -0,0 +1,1299 @@
...
+            y,
+            w,
+            h,
+            title,
+        }
+    }
+}
+
+pub struct Window {
+    window_socket: Arc<RedoxSocket>,
```
Comment (madsmtm) [inline src/platform_impl/redox/mod.rs:860]: Why are you not using `orbclient::Window`?

It seems like several things from that would be useful for us, both by using `orbclient::Window::events` in `EventLoop::run_return`, but also all the methods in `Window`?

**Review by madsmtm — CHANGES_REQUESTED** [review]
I had a look through the entire codebase, damn Orbital is _nice_!

Have a few suggestions, though take everything with a grain of salt, you're the expert in this domain ;)

File: src/platform_impl/redox/mod.rs (hunk 4 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:860]: orbclient::Window pairs the rendering context, which is wasteful.

File: src/platform_impl/redox/mod.rs [hunk 5]
```diff
@@ -0,0 +1,1299 @@
...
+        el: &EventLoopWindowTarget<T>,
+        attrs: window::WindowAttributes,
+        _: PlatformSpecificWindowBuilderAttributes,
+    ) -> Result<Self, error::OsError> {
+        let scale = MonitorHandle.scale_factor();
+
+        let (x, y) = if let Some(pos) = attrs.position {
+            pos.to_physical::<i32>(scale).into()
+        } else {
+            (-1, -1)
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:877]: (-1, -1) makes the window centered. Not the best thing, I realize, but that is what it is

File: src/platform_impl/redox/mod.rs [hunk 6]
```diff
@@ -0,0 +1,1299 @@
...
+};
+
+pub(self) use crate::platform_impl::Fullscreen;
+
+const ORBITAL_FLAG_ASYNC: char = 'a';
+const ORBITAL_FLAG_BACK: char = 'b';
+const ORBITAL_FLAG_FRONT: char = 'f';
+const ORBITAL_FLAG_BORDERLESS: char = 'l';
+const ORBITAL_FLAG_RESIZABLE: char = 'r';
+const ORBITAL_FLAG_TRANSPARENT: char = 't';
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:29]: No, they are not documented outside of the orbital source.

File: src/platform_impl/redox/mod.rs [hunk 7]
```diff
@@ -0,0 +1,1299 @@
...
+    }
+
+    pub fn raw_display_handle(&self) -> RawDisplayHandle {
+        RawDisplayHandle::Orbital(OrbitalDisplayHandle::empty())
+    }
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct WindowId {
+    fd: u64,
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:786]: impl From<u64> for WindowId is required for WindowId

File: src/platform_impl/redox/mod.rs [hunk 8]
```diff
@@ -0,0 +1,1299 @@
...
+
+const ORBITAL_FLAG_ASYNC: char = 'a';
+const ORBITAL_FLAG_BACK: char = 'b';
+const ORBITAL_FLAG_FRONT: char = 'f';
+const ORBITAL_FLAG_BORDERLESS: char = 'l';
+const ORBITAL_FLAG_RESIZABLE: char = 'r';
+const ORBITAL_FLAG_TRANSPARENT: char = 't';
+
+struct RedoxSocket {
+    fd: usize,
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:32]: The purpose of RedoxSocket is to allow read/write to be done on a const object, which File does not allow.

File: src/platform_impl/redox/mod.rs [hunk 9]
```diff
@@ -0,0 +1,1299 @@
...
+        let y = parts
+            .next()
+            .map_or(0, |part| part.parse::<i32>().unwrap_or(0));
+        let w = parts
+            .next()
+            .map_or(0, |part| part.parse::<u32>().unwrap_or(0));
+        let h = parts
+            .next()
+            .map_or(0, |part| part.parse::<u32>().unwrap_or(0));
+        let title = parts.next().unwrap_or("");
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:847]: Fixed

File: src/platform_impl/redox/mod.rs [hunk 10]
```diff
@@ -0,0 +1,1299 @@
...
+        }
+
+        //TODO: window_icon
+
+        // Open window
+        let window = unsafe {
+            RedoxSocket::open(&format!(
+                "orbital:{}/{}/{}/{}/{}/{}",
+                flag_str, x, y, w, h, attrs.title
+            ))
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:922]: Fixed

File: src/platform_impl/redox/mod.rs [hunk 11]
```diff
@@ -0,0 +1,1299 @@
...
+            let mut i = 0;
+            let mut resize_opt = None;
+            loop {
+                let window = {
+                    let windows = self.window_target.p.windows.read().unwrap();
+                    match windows.get(i) {
+                        Some(window) => window.clone(),
+                        None => break,
+                    }
+                };
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:390]: Fixed

File: src/platform_impl/redox/mod.rs (hunk 1 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:40]: Done

File: src/platform_impl/redox/mod.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (madsmtm) [reply src/platform_impl/redox/mod.rs:786]: Ah yeah, forgot that. No problem then

File: src/platform_impl/redox/mod.rs [hunk 12]
```diff
@@ -0,0 +1,1299 @@
...
+
+                let window_id = WindowId {
+                    fd: window.fd as u64,
+                };
+
+                let mut event_buf = [0u8; 16 * mem::size_of::<orbclient::Event>()];
+                let count =
+                    syscall::read(window.fd, &mut event_buf).expect("failed to read window events");
+                let events = unsafe {
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:406]: Done

File: src/platform_impl/redox/mod.rs [hunk 13]
```diff
@@ -0,0 +1,1299 @@
...
+                    let mut time = syscall::TimeSpec::default();
+                    timeout_socket.read(&mut time).unwrap();
+
+                    if let Some(duration) = instant.checked_duration_since(start) {
+                        time.tv_sec += duration.as_secs() as i64;
+                        time.tv_nsec += duration.subsec_nanos() as i32;
+                        while time.tv_nsec >= 1_000_000_000 {
+                            time.tv_sec += 1;
+                            time.tv_nsec -= 1_000_000_000;
+                        }
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:670]: When adding duration.subsec_nanos to time.tv_nsec, it could overflow. This resolves the overflow.

File: src/platform_impl/redox/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:29]: https://gitlab.redox-os.org/redox-os/orbital/-/blob/master/src/scheme.rs#L1377-1387

[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:29]: https://gitlab.redox-os.org/redox-os/orbital/-/blob/master/orbital-core/src/lib.rs#L538-547

File: src/platform_impl/redox/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (madsmtm) [reply src/platform_impl/redox/mod.rs:32]: Ah yeah, fair enough - what about `std::os::fd::OwnedFd` then?

File: src/platform_impl/redox/mod.rs (hunk 4 above)
Comment (madsmtm) [reply src/platform_impl/redox/mod.rs:878]: Hmm, yeah, that makes sense.

Still, one should perhaps consider deduplicating more of the code between `winit` and `orbclient`; but that can be done later.

File: src/platform_impl/redox/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (madsmtm) [reply src/platform_impl/redox/mod.rs:895]: Ok - would you mind documenting that ;)

File: src/platform_impl/redox/mod.rs [hunk 14]
```diff
@@ -0,0 +1,1299 @@
...
+                let events = unsafe {
+                    slice::from_raw_parts(
+                        event_buf.as_ptr() as *const orbclient::Event,
+                        count / mem::size_of::<orbclient::Event>(),
+                    )
+                };
+
+                for event in events {
+                    match event.to_option() {
+                        EventOption::Key(event) => {
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:417]: Done, but with all orbclient event types

File: src/platform_impl/redox/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (madsmtm) [reply src/platform_impl/redox/mod.rs:29]: Cool - would you mind adding these links to our source-code as well?

File: src/platform_impl/redox/mod.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (madsmtm) [reply src/platform_impl/redox/mod.rs:32]: Oh right sorry, that's outside our MSRV... Just forget it for now, then

File: src/platform_impl/redox/mod.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (madsmtm) [reply src/platform_impl/redox/mod.rs:672]: Ah right, `time` was non-null before the duration is added to it

File: src/platform_impl/redox/mod.rs [hunk 15]
```diff
@@ -0,0 +1,1299 @@
...
+impl<T> EventLoopProxy<T> {
+    pub fn send_event(&self, event: T) -> Result<(), event_loop::EventLoopClosed<T>> {
+        self.user_events_sender
+            .send(event)
+            .map_err(|mpsc::SendError(x)| event_loop::EventLoopClosed(x))?;
+
+        // Writing a default TimeSpec will always trigger a time event
+        self.wake_socket
+            .write(&syscall::TimeSpec::default())
+            .unwrap();
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:749]: I added a TimeSocket wrapper that I think cleans up most of the time handling

File: src/platform_impl/redox/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:917]: Done

File: src/platform_impl/redox/mod.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:34]: Done

**Review by madsmtm — APPROVED** [review]
Wonderful, thanks!

@kchibisov, @maroider and @msiglreith, do you approve that we add this additional platform? @jackpot51 will be maintaining it, but it will of course be a slight additional burden on us.

File: src/platform_impl/redox/mod.rs [hunk 16]
```diff
@@ -0,0 +1,1336 @@
...
+                i += 1;
+            }
+
+            while let Ok(event) = self.window_target.p.user_events_receiver.try_recv() {
+                event_handler(
+                    event::Event::UserEvent(event),
+                    &self.window_target,
+                    &mut control_flow,
+                );
+            }
```
Comment (kchibisov) [inline src/platform_impl/redox/mod.rs:636]: I think we deliver User events before other window events, but I'm not sure if it's stated anywhere.

File: src/platform_impl/redox/mod.rs [hunk 17]
```diff
@@ -0,0 +1,1336 @@
...
+                            };
+                        }
+                    }
+                } else {
+                    // Normal window event
+                    start_cause = StartCause::WaitCancelled {
+                        start,
+                        requested_resume,
+                    };
+                }
```
Comment (kchibisov) [inline src/platform_impl/redox/mod.rs:735]: You could use `match requested_resume` here with
```rust
Some(requested_resume) if event.id == timeout_socket.0.fd => {}
_ => {},
```

File: src/platform_impl/redox/mod.rs [hunk 18]
```diff
@@ -0,0 +1,1336 @@
...
+                ControlFlow::Poll => false,
+                ControlFlow::Wait => true,
+                ControlFlow::WaitUntil(instant) => {
+                    requested_resume = Some(instant);
+                    true
+                }
+                ControlFlow::ExitWithCode(code) => break code,
+            };
+
+            if wait {
```
Comment (kchibisov) [inline src/platform_impl/redox/mod.rs:673]: If can do `continue` from the match, no need for `wait` variable.

File: src/platform_impl/redox/mod.rs [hunk 19]
```diff
@@ -0,0 +1,1336 @@
...
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "orbital:{}/{}/{}/{}/{}/{}",
+            self.flags, self.x, self.y, self.w, self.h, self.title
+        )
+    }
+}
+
+pub struct Window {
```
Comment (kchibisov) [inline src/platform_impl/redox/mod.rs:901]: Could you split it into the separate file.

File: src/platform_impl/redox/mod.rs [hunk 20]
```diff
@@ -0,0 +1,1336 @@
...
+}
+
+#[derive(Clone, Copy, Debug, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct DeviceId;
+
+impl DeviceId {
+    pub const fn dummy() -> Self {
+        DeviceId
+    }
+}
```
Comment (kchibisov) [inline src/platform_impl/redox/mod.rs:848]: Should likely split it to. Would suggest to put it into `redox/mod.rs`, and then have `redox/window.rs` + `redox/event_loop.rs`.

File: src/platform_impl/redox/mod.rs [hunk 21]
```diff
@@ -0,0 +1,1336 @@
...
+    lctrl: bool,
+    rctrl: bool,
+    lalt: bool,
+    ralt: bool,
+    llogo: bool,
+    rlogo: bool,
+    left: bool,
+    middle: bool,
+    right: bool,
+}
```
Comment (kchibisov) [inline src/platform_impl/redox/mod.rs:221]: Could you use bitset with the help of `bitflag` crate.

**Review by kchibisov — CHANGES_REQUESTED** [review]
I can't test it, not that I wouldn't trust. But let's keep the code in a good shape from the start. Just by reading I got some understanding, but not sure if I've got everything covered with that review.

The major concern is that it's hard to follow due to everything being in one file and some functions being overly long.

Besides that it should be good. Also, could you while at it use `master` layout for the code, by that I mean that master struct comes before its fields declarations and it goes like that? That's subject, just find it easier to follow, since everything is below. And freestanding functions at the bottom of the file.

e.g.

```rust
struct TimeSocket(RedoxSocket);

impl TimeSocket {

}

impl Drop for TimeSocket {

}

impl OtherTrait for TimeSocket {
}

struct RedoxSocket;
```

File: src/platform_impl/redox/mod.rs (hunk 16 above)
Comment (madsmtm) [reply src/platform_impl/redox/mod.rs:636]: It's not really defined, see the top of `event.rs`.

On macOS, device and window events are delivered before user events

File: src/platform_impl/redox/mod.rs [hunk 22]
```diff
@@ -0,0 +1,1336 @@
...
+            if start_cause == StartCause::Init {
+                event_handler(
+                    event::Event::Resumed,
+                    &self.window_target,
+                    &mut control_flow,
+                );
+            }
+
+            // Handle window creates
+            while let Some(window) = self.window_target.p.creates.lock().unwrap().pop_front() {
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:371]: I will make it more explicit like the other loop.

File: src/platform_impl/redox/mod.rs [hunk 23]
```diff
@@ -0,0 +1,1336 @@
...
+                    },
+                    &self.window_target,
+                    &mut control_flow,
+                );
+
+                self.windows
+                    .retain(|window| window.fd as u64 != destroy_id.fd);
+            }
+
+            // Handle window events
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:418]: Done

File: src/platform_impl/redox/mod.rs (hunk 16 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:636]: I'm marking this as resolved, if macOS uses the same order

File: src/platform_impl/redox/mod.rs (hunk 17 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs]: Done

File: src/platform_impl/redox/mod.rs (hunk 20 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs:126]: Done

File: src/platform_impl/redox/mod.rs (hunk 19 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs]: Done

**Comment by jackpot51** [comment]
I rebased this to resolve conflicts

File: src/platform_impl/redox/mod.rs (hunk 18 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs]: Done

File: src/platform_impl/redox/mod.rs [hunk 24]
```diff
@@ -0,0 +1,1336 @@
...
+
+                // Send resize event on create to indicate first position
+                event_handler(
+                    event::Event::WindowEvent {
+                        window_id: window::WindowId(window_id),
+                        event: event::WindowEvent::Moved((properties.x, properties.y).into()),
+                    },
+                    &self.window_target,
+                    &mut control_flow,
+                );
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs]: It could be done in `Window::new`, but it would involve additional allocations to put these events into the user events channel

File: src/platform_impl/redox/mod.rs [hunk 25]
```diff
@@ -0,0 +1,1336 @@
...
+                                }
+                                event_handler(
+                                    #[allow(deprecated)]
+                                    event::Event::WindowEvent {
+                                        window_id: window::WindowId(window_id),
+                                        event: event::WindowEvent::KeyboardInput {
+                                            device_id: event::DeviceId(DeviceId),
+                                            input: event::KeyboardInput {
+                                                scancode: scancode as u32,
+                                                state: element_state(pressed),
```
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs]: Done

File: src/platform_impl/redox/mod.rs (hunk 21 above)
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs]: Done

File: src/platform_impl/redox/mod.rs (hunk 24 above)
[the remark this answers is not in the document]
Comment (jackpot51) [reply src/platform_impl/redox/mod.rs]: I'm going to say I vastly prefer the way it is right now compared to moving this to Window::new somehow. I could not simply use the user events channel since it only sends UserEvents right now. So I'd have to add more code in multiple places just to have these artificial events.

**Comment by jackpot51** [comment]
All requested changes have been applied or a reason provided for why they were not applied.

**Comment by jackpot51** [comment]
Due to the new platform abstraction I have renamed things to match, so Redox OS selects the orbital_platform config, and the implementation is in orbital modules instead of the previous redox modules. Theoretically, Orbital could be implemented by other Unix-like systems via a custom filesystem.

File: src/event.rs [hunk 26]
```diff
@@ -391,7 +391,7 @@ pub enum WindowEvent<'a> {
     ///
     /// ## Platform-specific
     ///
-    /// - **iOS / Android / Web:** Unsupported.
+    /// - **iOS / Android / Web / Redox OS:** Unsupported.
```
Comment (kchibisov) [inline src/event.rs:394]: I think using `Orbital` would be better? Maybe you'll have working Wayland platform at some point :p

File: src/platform/orbital.rs [hunk 27]
```diff
@@ -0,0 +1 @@
+// There are no Redox OS specific traits yet.
```
Comment (kchibisov) [inline src/platform/orbital.rs:1]: Should be likely `Orbital`?

File: src/platform_impl/orbital/event_loop.rs [hunk 28]
```diff
@@ -0,0 +1,708 @@
...
+    ButtonEvent, EventOption, FocusEvent, HoverEvent, KeyEvent, MouseEvent, MoveEvent, QuitEvent,
+    ResizeEvent, ScrollEvent, TextInputEvent,
+};
+use raw_window_handle::{OrbitalDisplayHandle, RawDisplayHandle};
+use std::{
+    collections::VecDeque,
+    mem, slice,
+    sync::{mpsc, Arc, Mutex},
+    time::Instant,
+};
```
Comment (kchibisov) [inline src/platform_impl/orbital/event_loop.rs:11]: Could you put `std` import as a dedicated import at the top? And then after a blank like 3d party crates?

**Review by kchibisov — DISMISSED** [review]
Just one minor note, should be good to go otherwise.

File: src/event.rs (hunk 26 above)
Comment (jackpot51) [reply src/event.rs:394]: Yep, good point

**Comment by madsmtm** [comment]
I've added you to the @rust-windowing/winit team, so you should have write access to the repo now.

Feel free to merge any PR that only/primarily touches Orbital, as well as minor PRs, without anyone else's input - for the rest, just use your good judgement, of which I'm sure you have plenty ;)

**Comment by Snowiiii** [comment]
gg

File: src/platform/orbital.rs (hunk 27 above)
Comment (jackpot51) [reply src/platform/orbital.rs:1]: Done

File: src/platform_impl/orbital/event_loop.rs (hunk 28 above)
Comment (jackpot51) [reply src/platform_impl/orbital/event_loop.rs:6]: Done
